packages/typespec-client-generator-core/src/decorators.ts

```diff
@@ -956,19 +956,78 @@ const alternateTypeKey = createStateSymbol("alternateType");
 export const $alternateType: AlternateTypeDecorator = (
   context: DecoratorContext,
   source: ModelProperty | Scalar,
-  alternate: Scalar,
+  alternate: Type,
   scope?: LanguageScopes,
 ) => {
-  if (source.kind === "ModelProperty" && source.type.kind !== "Scalar") {
+  let invalidAlternate = true;
+
+  if (alternate.kind === "Intrinsic" && alternate.name === "unknown") {
+    invalidAlternate = false;
+  }
+
+  if (
+    alternate.kind === "Model" &&
+    alternate.name === "Array" &&
+    alternate.indexer?.value.kind === "Scalar"
+  ) {
+    invalidAlternate = false;
+  }
+
+  if (alternate.kind === "Scalar") {
+    invalidAlternate = false;
+  }
+
+  if (invalidAlternate) {
+    reportDiagnostic(context.program, {
+      code: "invalid-alternate-type",
+      format: {},
+      target: alternate,
+    });
+    return;
+  }
+
+  let invalidSource = true;
+
+  if (alternate.kind === "Intrinsic" && alternate.name === "unknown") {
+    if (
+      (source.kind === "ModelProperty" && source.type.kind === "Scalar") ||
+      source.kind === "Scalar"
+    ) {
+      invalidSource = false;
+    }
+  }
+
+  if (alternate.kind === "Scalar") {
+    if (
+      (source.kind === "ModelProperty" && source.type.kind === "Scalar") ||
+      source.kind === "Scalar"
+    ) {
+      invalidSource = false;
+    }
+  }
+
+  if (alternate.kind === "Model") {
+    if (
+      source.kind === "ModelProperty" &&
+      source.type.kind === "Model" &&
+      source.type.name === "Array" &&
+      source.type.indexer?.value.kind === "Scalar"
+    ) {
+      invalidSource = false;
+    }
+  }
+
+  if (invalidSource) {
     reportDiagnostic(context.program, {
       code: "invalid-alternate-source-type",
```

Comment:
so do we care if the alternate type is not compatible with the type of the source? Such like I think in our current implementation, it would not fail if we do:
```
model Model {
	@alternateType(AzureLocation)
	prop: string[];
```
Is it a goal that we find such cases like this?

Reply:
we do have the check for it, if the alternate type is scalar then the source type must also be scalar.